src/v/archival/archiver_scheduler_api.h

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "base/outcome.h"
+#include "model/fundamental.h"
+#include "model/ktp.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/future.hh>
+
+namespace ss = seastar;
+
+namespace archival {
+
+/// The object is responsible of managing execution of different
+/// workflows. It controls backoff in situation when there is no
+/// new data to upload. It can also apply throttling so one workflow
+/// wouldn't be able to use too many requests.
+/// It also acts as a scheduler that organizes interleaved execution
+/// of different workflow that belong to the same NTP by guaranteeing
+/// mutual exclusion. Only one workflow per-NTP is allowed to be active
+/// at any moment in time. To guarantee this every workflow should start
+/// by calling `maybe_suspend_*` and should also invoke `maybe_suspend_*`
+/// periodically to allow interleaving.
+/// Suspend methods are workflow specific.
+class archiver_scheduler_api {
+public:
+    archiver_scheduler_api() = default;
+    archiver_scheduler_api(const archiver_scheduler_api&) = delete;
+    archiver_scheduler_api(archiver_scheduler_api&&) noexcept = delete;
+    archiver_scheduler_api& operator=(const archiver_scheduler_api&) = delete;
+    archiver_scheduler_api& operator=(archiver_scheduler_api&&) noexcept
+      = delete;
+    virtual ~archiver_scheduler_api();
+
+    struct suspend_upload_arg {
+        model::ktp ntp;
+        // Set to true if the manifest was uploaded, false if not, nullopt if
+        // there is no change from the previous state
+        std::optional<bool> manifest_dirty{std::nullopt};
```

Comment:
the archival STM has 'dirty' flag which works the same way, so this just replicates this convention
from STM perspective the manifest is clean if we replicated the batch that adjust the clean offset. In this case the manifest is not dirty because STM matches state in the cloud. If the manifest was uploaded but it doesn't match the state of the STM then the manifest is dirty (the 'manifest' is the stm manifest in this case).